utils/cpuload/netlink.go

```diff
@@ -0,0 +1,232 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cpuload
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"syscall"
+)
+
+const (
+	GENL_ID_CTRL            = syscall.NLMSG_MIN_TYPE
+	GENL_VERSION_CTRL       = 0x1
+	TASKSTATS_GENL_NAME     = "TASKSTATS"
+	CGROUPSTATS_CMD_ATTR_FD = 0x1
+	CTRL_ATTR_FAMILY_ID     = 0x1
+	CTRL_ATTR_FAMILY_NAME   = 0x2
+	CTRL_CMD_GETFAMILY      = 0x3
+)
+
+type GenMsghdr struct {
+	Command  uint8
+	Version  uint8
+	Reserved uint16
+}
+
+type NetlinkMessage struct {
+	Header    syscall.NlMsghdr
+	GenHeader GenMsghdr
+	Data      []byte
+}
+
+func (self NetlinkMessage) toRawMsg() (rawmsg syscall.NetlinkMessage) {
+	rawmsg.Header = self.Header
+	w := bytes.NewBuffer([]byte{})
+	binary.Write(w, binary.LittleEndian, self.GenHeader)
+	w.Write(self.Data)
+	rawmsg.Data = w.Bytes()
+	return rawmsg
+}
+
+type LoadStatsResp struct {
+	Header    syscall.NlMsghdr
+	GenHeader GenMsghdr
+	Stats     LoadStats
+}
+
+// Return required padding to align 'size' to 'alignment'.
+func padding(size int, alignment int) int {
+	unalignedPart := size % alignment
+	return (alignment - unalignedPart) % alignment
+}
+
+// Get family id for taskstats subsystem.
+func getFamilyId(conn *Connection) (uint16, error) {
+	msg := prepareFamilyMessage()
+	conn.WriteMessage(msg.toRawMsg())
+
+	resp, _ := conn.ReadMessage()
+	id, err := parseFamilyResp(resp)
+	if err != nil {
+		return 0, err
+	}
+	return id, nil
+}
+
+// Append an attribute to the message.
+// Adds attribute info (length and type), followed by the data and necessary padding.
+// Can be called multiple times to add attributes. Only fixed size and string type
+// attributes are handled. We don't need nested attributes for task stats.
+func addAttribute(buf *bytes.Buffer, attrType uint16, data interface{}, dataSize int) {
+	attr := syscall.RtAttr{
+		Len:  syscall.SizeofRtAttr,
+		Type: attrType,
+	}
+	attr.Len += uint16(dataSize)
+	binary.Write(buf, binary.LittleEndian, attr)
```

Comment:
Moved the endianness to a global var for now. I need to test what it does on big-endian machines, but I think the kernel might be reporting these in a specific order. We can always check the machine and set up endianness globally if that's not the case. Added a TODO.